src/server/service/tracing.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::Arc;
+// use std::lazy::Lazy;
+use std::time::Duration;
+
+use arc_swap::ArcSwap;
+use chrono::{DateTime, Local};
+use futures::channel::mpsc::{channel, Receiver, Sender};
+use futures::prelude::*;
+use kvproto::tracepb;
+use minitrace::prelude::*;
+use online_config::{ConfigChange, OnlineConfig};
+use serde_json::json;
+use std::time::UNIX_EPOCH;
+use tikv_util::time::{duration_to_sec, Instant};
+
+use crate::server::metrics::{GrpcTypeKind, GRPC_MSG_HISTOGRAM_STATIC};
+
+type CollectTask = (Collector, Vec<tracepb::RemoteParentSpan>, Duration);
+
+const CHANNEL_SIZE: usize = 1024;
+
+pub fn init_tracing(
+    config: &TracingConfig,
+) -> (TracingService, TracerFactory, TracingConfigManager) {
+    let config = Arc::new(ArcSwap::from_pointee(config.clone()));
+    let (tx, rx) = channel(CHANNEL_SIZE);
+
+    (
+        TracingService { rx },
+        TracerFactory {
+            config: config.clone(),
+            tx,
+        },
+        TracingConfigManager { config },
+    )
+}
+
+pub struct TracingService {
+    rx: Receiver<CollectTask>,
+}
+
+impl TracingService {
+    // TODO (andylokandy): implement a real pubsub service
+    pub fn register_service(self) {
+        std::thread::spawn(move || {
+            let print_to_slow_log = self.rx.for_each_concurrent(
+                None,
+                |(collector, remote_parent_spans, total_duration)| async move {
+                    let mut report = TracingReport {
+                        spans: collector.collect().await,
+                        remote_parent_spans,
+                    };
+                    warn!(
+                        "[tracing] slow request deteted";
+                        "total_duration" => ?total_duration,
+                        "report" => serde_json::to_string(&report.to_json()).unwrap_or_default()
+                    );
+                },
+            );
+            futures::executor::block_on(print_to_slow_log);
+        });
+    }
+}
+
+#[macro_export]
+macro_rules! new_request_tracer {
```

Comment:
Is there any special reason that we need to manually take fields in `TracerFactory` and pass them to `new_tracer` instead of letting `new_tracer` just take `&self`?

Reply:
It's to avoid manually writing duplicated `crate::server::metrics::GrpcTypeKind::$req_name` and `"$req_name"`.